### PR TITLE
CORDA-1845: Check for min plaform version of 4 when building transactions with reference states

### PR DESCRIPTION
**Pull request by shamsasari**
Also includes some minor cleanup brought up in a previous PR.

### CODE REVIEW

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt [hunk 1]
```diff
@@ -95,11 +96,23 @@ open class TransactionBuilder @JvmOverloads constructor(
      * [HashAttachmentConstraint].
      *
      * @returns A new [WireTransaction] that will be unaffected by further changes to this [TransactionBuilder].
+     *
+     * @throws IllegalStateException if there are reference states and the minimum platform version is less than 4.
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: Can we define an unchecked custom exception for zone version being too low?

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt [hunk 2]
```diff
@@ -95,11 +96,23 @@ open class TransactionBuilder @JvmOverloads constructor(
...
      */
     @Throws(MissingContractAttachments::class)
     fun toWireTransaction(services: ServicesForResolution): WireTransaction = toWireTransactionWithContext(services)
 
+    @CordaInternal
     internal fun toWireTransactionWithContext(services: ServicesForResolution, serializationContext: SerializationContext? = null): WireTransaction {
+        val referenceStates = referenceStates()
+        if (referenceStates.isNotEmpty()) {
+            val minimumPlatformVersion = services.networkParameters.minimumPlatformVersion
+            check(minimumPlatformVersion >= 4) {
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: Can we extract this out to a function that throws a custom exception, so we always get the useful error message for every case where we do this type of check in future?

File: core/src/test/kotlin/net/corda/core/transactions/TransactionBuilderTest.kt [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import org.junit.Rule
+import org.junit.Test
+
+class TransactionBuilderTest {
+    @Rule
+    @JvmField
+    val testSerialization = SerializationEnvironmentRule()
+
+    private val notary = TestIdentity(DUMMY_NOTARY_NAME).party
+    private val services = rigorousMock<ServicesForResolution>()
```
Comment (mikehearn) [inline core/src/test/kotlin/net/corda/core/transactions/TransactionBuilderTest.kt]: Mockito can be quite slow, maybe we can introduce a custom mock for ServicesForResolution for this sort of case of testing transaction building?

Comment (shamsasari) [reply core/src/test/kotlin/net/corda/core/transactions/TransactionBuilderTest.kt]: I think something for a separate PR.

**Comment by shamsasari** [comment]
Comments addressed.

**Pull request merged**